### ISSUE TITLE
Mirror amp macros in RTC call back in response

### ISSUE THREAD

**Issue by patmmccann** [prebid/prebid-server#1167]
Several values are available as AMP macros that it is not currently possible to use as set targeting values. 

PAGEVIEWID - pageViewId
HREF - equivalent to window.context.location.href
ATTR(height) - Height attribute of the amp-ad element
ATTR(width) - Width attribute of the amp-ad element
ATTR(data-slot) - data-slot attribute of the amp-ad element
ATTR(data-multi-size) - data-multi-size attribute of the amp-ad element
ATTR(data-multi-size-validation) - data-multi-size-validation attribute of the amp-ad element
ATTR(data-override-width) - data-override-width attribute of the amp-ad element
ATTR(data-override-height) - data-override-height attribute of the amp-ad element
ATTR(data-json) - data-json attribute of the amp-ad element
ADCID - adClientId
TGT - Just the targeting piece of data-json
CANONICAL_URL - The canonical URL of the page.
TIMEOUT - The publisher-specified timeout for the RTC callout.

A possible strategy to set targeting to gam with one of these values is to pass one of them to prebid server and have prebid server include it in the response. 

What we would like to do is pass a pageviewid to gam so that we can associate impressions from the same pageview together on our gam log.

**Comment by bretg** [comment]
We discussed this in the Prebid Server meeting today. The suggested approach is:

1) PBS parses the AMP query string variables and puts them all in ext.prebid.amp.data

2) The AMP Stored Request (or anything else that creates PBS openRTB!) could specify which adserver targeting variables to add and their values in a generic way. Proposed:

```
   ext.prebid.adservertargeting:  [{
              "key": "hb_ampparam",
              "source": "bidrequest",    // causes PBS to treat 'value' as a path to pull from the request object
              "value": "ext.prebid.amp.data.ampkey"
            },{
               "key": "hb_static_thing",
               "source": "static",  // causes PBS to just use the 'value' provided
               "value": "my-static-value"
            },{
               "key": "{{BIDDER}}_custom1", // {{BIDDER}} is a macro to be resolved
               "source": "bidresponse",  // causes PBS to treat 'value' as a path to pull from the bidder's response object, specifically seatbid[j].bid[k]
               "value": "seatbid.bid.ext.custom1"
            }
   }]
```
3) While generating the response seatbid[].bid[].ext.prebid.targeting, Prebid Server would parse the request ext.prebid.adservertargeting and follow these rules:

a) if source=bidrequest, then use the path specified in 'value' to look for data in the ortb bidrequest. e.g. ext.prebid.amp.data.ampkey. If imp is specified in the path, the assumption is that the specific imp[] desired corresponds to the seatbid[].bid[] we're working on. i.e. imp[].id=seatbid[].bid[].impid. For example the path could be imp.ext.customthing

b) if source==bidresponse, then use the path specified in 'value' to look for data in the ortb bidresponse from the bidder. e.g. seatbid.bid.ext.custom1. The reference to `seatbid.bid` is assumed to line up with the current seatbid.bid object we're adding targets to.

c) if source==static, just copy the 'value' straight up

d) if the 'key' contains {{BIDDER}}, it's treated as a macro to be resolved with the biddercode from the current seatbid.seat.

**Comment by SyntaxNode** [comment]
Do we also want to add `ATTR(data-multi-size-validation)` and `ATTR(data-json)` as well? They aren't in the params list today.

**Comment by hhhjort** [comment]
I would suggest we modify Brett's proposed solution to defining the custom keys from the AMP query string to look something like:
`ext.prebid.adservertargeting:  [{
              "key": "hb_ampparam",
              "source": "querystring",
              "value": "ampkey"
            }]`

Embedding the syntax in the request object implies we will load a copy of the amp query string parameters into the request JSON. Either we don't actually modify the request JSON this way, which could potentially confuse adapter authors, or we go through all that work on every case to support whatever corner case an adapter author can come up with for accessing that data. Given that even with this inclusion, the adapter cannot determine the original value in the request before it was overwritten by the querystring, nor even if it was set to begin with, I doubt there exists a case to include the query string values.

**Comment by SyntaxNode** [comment]
Thoughts? @bretg @patmmccann

**Comment by bretg** [comment]
I prefer the original proposal we discussed in the committee meeting:

1) it's more generic, supporting future non-AMP, non-querystring interfaces
2) it's more consistent that Prebid Server core operates only the OpenRTB 'object of record'. I view the AMP 'module' to have just one job: convert the AMP query string to our canonical form.
2) copying AMP query string params to ext.prebid.amp.data is just a way to normalize the incoming request to our 'object of record', though I agree that adapters shouldn't have access to it. They only get data via the [first party data](https://docs.google.com/document/d/1PPCojtui1WXkNQdiydqYGZ3i_66Uxi7JlrcpQT9cGus/edit) path. I've updated the proposal above to note removing ext.prebid.amp.data before sending to adapters.

**Comment by bretg** [comment]
FYI - done in PBS-Java 1.41

**Comment by bretg** [comment]
This has been documented at http://docs.prebid.org/prebid-server/endpoints/openrtb2/pbs-endpoint-auction.html#custom-targeting-pbs-java-only

**Comment by bretg** [comment]
@patmmccann - PAGEVIEWID wasn't part of the standard AMP callout-vendors. Opened https://github.com/ampproject/amphtml/pull/30712 to add pvid for the final piece of this request:

```
What we would like to do is pass a pageviewid to gam so that we can associate impressions from the same pageview together on our gam log.
```

**Comment by VeronikaSolovei9** [comment]
@bretg Could you please confirm where final targeting data should be placed in response? Is it `seatbid[].bid[].ext.prebid.targeting`?

**Comment by bretg** [comment]
Hi @VeronikaSolovei9 -The targeting location has always been seatbid[].bid[].ext.prebid.targeting -- is it mis-documented somewhere. Seems ok in this ticket.

**Comment by VeronikaSolovei9** [comment]
Got it, thank you Bret! Also another question. If data with key  f.i.`hb_ampparam` exists already in `seatbid[].bid[].ext.prebid.targeting` and there is value with same name comes from  ad server targeting, which one takes precedence? Should we overwrite existing values with values from ad server targeting?

**Comment by bretg** [comment]
Not clear how that could happen in the PBS world. Let's say that the stuff in ext.prebid.adservertargeting takes precedence should there be a collision.

**Comment by VeronikaSolovei9** [comment]
Bret, thank you for the clarification.
I have some more questions.
1. In case targeting config points to "bidresponse" field itself  - is this a valid case: 
```
{
       "key": "custom1", 
       "source": "bidresponse",  
       "value": "seatbid.cur"
            }
```
Bidder response has limited number of properties that are reasonable to use as a targeting data, except of `seatbid` and `ext`. Do we still want to support this case, f.i. `seatbid.cur`? 
In [this document,](https://iabtechlab.com/wp-content/uploads/2022/04/OpenRTB-2-6_FINAL.pdf) please refer to paragraph 4.3.1 (format is the same for OpenRTB 2.5) - it describes fields in bidresponse.

2. In case error occured during targeting keys processing - should the entire response fail? If not should "successful" targeting keys be included to response or omit all targeting keys? Should the warning be returned? 

3. Is targeting key always a string? Can key point to a struct. f.i. "seatbid.bid.cat" is `"[\"cat1\",\"cat2\"]"`

**Comment by bretg** [comment]
> 1. In case targeting config points to "bidresponse" field itself - is this a valid case: seatbid.cur

Sure.

> 2. In case error occured during targeting keys processing - should the entire response fail? 

No. It should add the targets it can and emit a debug warning for any targets that can't be processed.

> 3.  Can key point to a struct.

Good point. I verified how PBS-Java is handling this -- simple values only: ints and strings. It ignores objects and arrays. Will update the description.

**Comment by VeronikaSolovei9** [comment]
Got it, thank you Bret!

**Comment by SyntaxNode** [comment]
Implemented in PBS-Go 0.249.0.
